content/en/docs/reference/using-api/deprecation-policy.md

```diff
@@ -365,6 +365,40 @@ This applies only to significant, user-visible behaviors which impact the
 correctness of applications running on Kubernetes or that impact the
 administration of Kubernetes clusters, and which are being removed entirely.
 
+The exception to the above rule is **feature-gates**. Feature gates are key=value 
+pairs that enable users to enable/disable experimental features.
+Features can be in Alpha, Beta or GA stage.
+
+Feature gates are intended to cover the development of a feature - they are not 
+intended to be long-term APIs. As such, they are expected to be deprecated and 
+removed after a feature becomes GA or is dropped. 
+As a user of feature gates pay atention to deprecation announcements for feature-gates 
```

Comment:
to illustrate the feature gate as a toggle for a feature that is GA, toggling out a feature that is being deprecated.

the "core-dns" / "kube-dns" toggle scenario:
- core-dns is GA and the feature gate is `true` by default.
- if the feature gate is set to `false` it falls back to kube-dns.
- kube-dns might take a while before being deprecated.
- the removal of kube-dns will trigger the removal of the feature gate.

if we choose to move kube-dns outside of the feature gate toggle:
- we probably have to introduce a new CLI flag called `--enable-kube-dns` next to the `--feature-gates=CoreDNS=...`
- this introduces a flag that has to exist until kube-dns is removed and it has to follow a deprecation policy too.
- the flag and the feature gate have to be mutually exclusive.

Reply:
"Deprecated feature-gates for GA features are still accepted (i.e. do not cause errors) but do not have any effect." -- I believe we wanted to express that feature gates for GA features are discouraged, however if we say that they exist but do not not toggle features I am ok with that. :) 
@thockin , could you please elaborate on: 
"Features that are intended to be user-disabled should include mechanisms for that, which are themselves covered by the feature-gate."
Is this a negation of the sentence above it?

Reply:
@neolit123 , yes! :) I tried to describe that case in the following sentence, now I am thinking of a way to keep it but make it a bit less verbose. Any suggestions are welcome :) 
"The feature is a candidate to become a configurable option, this happens when the old behavior is needed by users. The new option is introduced as soon as we identify the need and it follows the feature lifecycle as well as feature deprecation rules."

Reply:
I don't think the feature gate should control selecting which DNS to use.  E.g. for this particular case, wouldn't it be cleaner to do something like:

Assume the feature is alpha.

Add a `--dns-provider=[kubedns | coredns]` flag.  If the feature gate is false, the only value allowed is kubedns.  If the feature-gate is true, coredns is acceptable.

Now the feature goes beta.

Same behavior.

Now the feature goes GA.  The gate no longer disables the option to use coredns, but it is still an option.

At this point you can deprecate support for kubedns (or not) and change the default to coredns

CoreDNS is a weird example because we GA'ed it before making it the default, but in this case it works - you just hold off deprecating the kubedns support.  The main point is that there is a distinction between the feature going GA and removing the ability to choose.  Most features do not need an ability to choose, once they go GA.

I am mostly arguing that (for MOST features) we don't want to support disabling them once they go GA.  But as I said, I could argue this both ways.  I am trying to reduce the number of conditional behaviors here.  Someone who feels I am introducing risk should argue with me...

Reply:
> Add a --dns-provider=[kubedns | coredns] flag. If the feature gate is false, the only value allowed is kubedns. If the feature-gate is true, coredns is acceptable.

i guess this rolls back to my previous point about:

> - this introduces a flag that has to exist until kube-dns is removed and it has to follow a deprecation policy too.

once kube-dns is removed as a feature, the `--dns-provider` flag might get removed too, unless it's left there as a no-op, because coredns would be the only option at that point...yet more options might be added later, which makes the flag relevant again and probably should never be removed.

> CoreDNS is a weird example because we GA'ed it before making it the default, but in this case it works - you just hold off deprecating the kubedns support. The main point is that there is a distinction between the feature going GA and removing the ability to choose. Most features do not need an ability to choose, once they go GA.

if we can treat coredns vs kubedns as a corner case, it's possibly best to treat the doc with a recommendation to always remove a gate once GA without outlining exceptions, although leaving the degree of freedom should not discarded.